text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+
+# The C++ Solution
+
+C++'s [P1478] proposes the addition of these two functions to the C++ standard
+library to solve this problem:
+
+```cpp
+void atomic_load_per_byte_memcpy(void *dest, void *source, size_t, memory_order);
+void atomic_store_per_byte_memcpy(void *dest, void *source, size_t, memory_order);
+```
+
+The first one is effectively a series of `AtomicU8::load`s followed by a memory fence,
+while the second one is basically a memory fence followed by series of `AtomicU8::store`s.
+Except the implementation can be much more efficient.
+The implementation is allowed to load/store the bytes in any order,
+and doesn't have to operate on individual bytes.
+
+The memory order can only be relaxed, acquire (for load), and release (for store).
+Sequentially consistent ordering for these operations is disallowed,
+since it's not obvious what that means for these tearable operations.
+
+# The Rust Solution
+
+While C++'s solution can be easily copy-pasted into Rust with a nearly identical signature,
+it wouldn't fit with the rest of our atomic APIs.
+
+All our atomic operations happen through the `Atomic*` types,
+and we don't have atomic operations that operate on raw pointers.
+(Other than as unstable intrinsics.)
+
+Adding this functionality as a variant on `copy_nonatomic`, similar to the C++ solution,
+would not be very ergonomic an can easily result in subtle bugs causing undefined behavior.
+
+Instead, I propose to add a `AtomicPerByte<T>` type
+similar to our existing atomic types: a `Sync` storage for a `T`
+that can be written to and read from by multiple threads concurrently.
+
+The `SeqLock` implementation above would use this type instead of an `UnsafeCell`.
+It'd no longer need an unsafe `Sync` implementation,
+since the `AtomicPerByte<T>` type can be shared between threads safely.
+
+This type has a (safe!) `store` method consuming a `T`,
+and a (safe!) `load` method producing a `MaybeUninit<T>`.
+The `MaybeUninit` type is used to represent the potentially invalid state
+the data might be in, since it might be the result of tearing during a race.
+
+Only after confirming that there was no race and the data is valid
+can one safely use `MaybeUninit::assert_init` to get the actual `T` out.
+
+# Full API Overview
+
+The `AtomicPerByte<T>` type can be thought of as
+the `Sync` (data race free) equivalent of `MaybeUninit<T>`.
+It can contain a `T`, but it might be invalid in various ways
+due to concurrent store operations.
+Its interface resembles a mix of the interfaces of `MaybeUninit` and the atomic types.
+
+```rust
+#[repr(transparent)]
+struct AtomicPerByte<T> { inner: UnsafeCell<MaybeUninit<T>> }
+
+unsafe impl<T: Send> Sync for AtomicPerByte<T> {}
+
+impl<T> AtomicPerByte<T> {
+    pub const fn new(value: T) -> Self;
+    pub const fn uninit() -> Self;
+
+    pub fn load(&self, ordering: Ordering) -> MaybeUninit<T>;
+    pub fn store(&self, value: T, ordering: Ordering);
```

Comment:
That's good to know, thanks.

For this specific API: instead of `store`, `store_from` can be used to store something from a `&MaybeUninit<T>`. I figured the majority of the cases would use a `T`, so that's what I picked for the `store` method, to keep it as simple as possible. (There doesn't seem to be much use to consuming a `MaybeUninit<T>` since it doesn't have a drop implementation anyway.)